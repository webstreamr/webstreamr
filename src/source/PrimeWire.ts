import crypto from 'crypto';
import * as cheerio from 'cheerio';
import { JSDOM } from 'jsdom';
import { ContentType } from 'stremio-addon-sdk';
import { Context, CountryCode } from '../types';
import { Fetcher, getImdbId, Id, ImdbId } from '../utils';
import { Source, SourceResult } from './types';

export class PrimeWire implements Source {
  public readonly id = 'primewire';

  public readonly label = 'PrimeWire';

  public readonly contentTypes: ContentType[] = ['movie', 'series'];

  public readonly countryCodes: CountryCode[] = [CountryCode.en];

  private readonly baseUrl = 'https://www.primewire.tf';

  private readonly fetcher: Fetcher;

  private readonly redirectUrlCache = new Map<string, URL>();

  public constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  public async handle(ctx: Context, _type: string, id: Id): Promise<SourceResult[]> {
    const imdbId = await getImdbId(ctx, this.fetcher, id);

    // We need any non-protected entrypoint to grab the app.js
    const html = await this.fetcher.text(ctx, new URL('/legal', this.baseUrl));
    const appJsMatch = html.match(/src="(.*?app-.*\.js.*?)"/) as string[];
    const appJs = await this.fetcher.text(ctx, new URL(appJsMatch[1] as string, this.baseUrl));

    const pageUrl = await this.fetchPageUrl(ctx, appJs, imdbId);
    if (!pageUrl) {
      return [];
    }

    const pageHtml = await this.fetcher.text(ctx, pageUrl);

    let linksHtml = pageHtml;
    if (imdbId.season) {
      const episodeUrl = await this.fetchEpisodeUrl(pageHtml, imdbId);
      if (!episodeUrl) {
        return [];
      }

      linksHtml = await this.fetcher.text(ctx, episodeUrl);
    }

    // Yes, we are aware of the risks.. ðŸ˜°
    const jsdom = new JSDOM(linksHtml, { runScripts: 'dangerously' });
    const scriptElement = jsdom.window.document.createElement('script');
    scriptElement.textContent = appJs;
    jsdom.window.document.body.appendChild(scriptElement);

    const $ = cheerio.load(jsdom.window.document.documentElement.outerHTML);

    return Promise.all(
      $(`a.go-link`)
        .map((_i, el) => new URL($(el).attr('href') as string, this.baseUrl))
        .toArray()
        .map(async (redirectUrl) => {
          let targetUrl = this.redirectUrlCache.get(redirectUrl.href);

          if (!targetUrl) {
            const headers = await this.fetcher.head(ctx, redirectUrl, { redirect: 'manual' });

            targetUrl = new URL(headers['location'] as string);
            this.redirectUrlCache.set(redirectUrl.href, targetUrl);
          }

          return { countryCode: CountryCode.en, url: targetUrl };
        }),
    );
  };

  private readonly fetchPageUrl = async (ctx: Context, appJs: string, imdbId: ImdbId): Promise<URL | undefined> => {
    const sha1SuffixMatch = appJs.match(/s\.value\+"(.*?)"/) as string[];

    const ds = crypto.createHash('sha1')
      .update(imdbId.id + sha1SuffixMatch[1])
      .digest('hex')
      .slice(0, 10);

    const searchResults = await this.fetcher.text(ctx, new URL(`/filter?s=${encodeURIComponent(imdbId.id)}&ds=${encodeURIComponent(ds)}`, this.baseUrl));

    const $ = cheerio.load(searchResults);

    return $('.index_item a')
      .map((_i, el) => new URL($(el).attr('href') as string, this.baseUrl))
      .get(0);
  };

  private readonly fetchEpisodeUrl = async (pageHtml: string, imdbId: ImdbId): Promise<URL | undefined> => {
    const $ = cheerio.load(pageHtml);

    return $(`div[data-id="${imdbId.season}"]`)
      .children(`.tv_episode_item:nth-child(${imdbId.episode})`)
      .children('a[href]')
      .map((_i, el) => new URL($(el).attr('href') as string, this.baseUrl))
      .get(0);
  };
}
