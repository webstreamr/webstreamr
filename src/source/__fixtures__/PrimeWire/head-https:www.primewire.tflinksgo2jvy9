{"alt-svc":"h3=\":443\"; ma=86400","cache-control":"max-age=0, private, must-revalidate","cf-cache-status":"DYNAMIC","cf-ray":"95fc473f4e599188-FRA","connection":"close","content-type":"text/html; charset=utf-8","cross-origin-window-policy":"deny","date":"Tue, 15 Jul 2025 21:15:59 GMT","location":"https://bigwarp.cc/4tb37obzt6re","nel":"{\"report_to\":\"cf-nel\",\"success_fraction\":0.0,\"max_age\":604800}","report-to":"{\"group\":\"cf-nel\",\"max_age\":604800,\"endpoints\":[{\"url\":\"https://a.nel.cloudflare.com/report/v4?s=5gyYWzhG6RoXPZadikiDNxf2LVB4GTRf97%2Fq7l%2F4rPs%2Bw%2BJ7XDI1wWTDBCKe6fksZp%2FLgiisRcgK7laoz04XleXiRsaMJxIlIuK89uuUTbg%3D\"}]}","server":"cloudflare","x-content-type-options":"nosniff","x-download-options":"noopen","x-frame-options":"SAMEORIGIN","x-permitted-cross-domain-policies":"none","x-request-id":"GFKJSsiXd_8HIvwJH_uH","x-xss-protection":"1; mode=block"}