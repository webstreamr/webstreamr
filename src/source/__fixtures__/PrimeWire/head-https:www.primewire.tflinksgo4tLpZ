{"alt-svc":"h3=\":443\"; ma=86400","cache-control":"max-age=0, private, must-revalidate","cf-cache-status":"DYNAMIC","cf-ray":"95fc467cae98d2af-FRA","connection":"close","content-type":"text/html; charset=utf-8","cross-origin-window-policy":"deny","date":"Tue, 15 Jul 2025 21:15:28 GMT","location":"https://voe.sx/grsvpdbxvw3e","nel":"{\"report_to\":\"cf-nel\",\"success_fraction\":0.0,\"max_age\":604800}","report-to":"{\"group\":\"cf-nel\",\"max_age\":604800,\"endpoints\":[{\"url\":\"https://a.nel.cloudflare.com/report/v4?s=Oei9GlhSFoCfGjGdjbIlEtZuGbefBQnI2SpMb%2FQKmFdSOeZ4l7IHkRM4psu8sRztZDJrnhBwj7yNQ2gNt443luKElavwCZJLLt66KnBXuRg%3D\"}]}","server":"cloudflare","x-content-type-options":"nosniff","x-download-options":"noopen","x-frame-options":"SAMEORIGIN","x-permitted-cross-domain-policies":"none","x-request-id":"GFKJQ4fXHdu7En4Qy1eD","x-xss-protection":"1; mode=block"}