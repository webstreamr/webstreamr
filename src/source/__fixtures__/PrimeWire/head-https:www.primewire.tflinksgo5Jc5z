{"alt-svc":"h3=\":443\"; ma=86400","cache-control":"max-age=0, private, must-revalidate","cf-cache-status":"DYNAMIC","cf-ray":"95fc46a46e001cab-FRA","connection":"close","content-type":"text/html; charset=utf-8","cross-origin-window-policy":"deny","date":"Tue, 15 Jul 2025 21:15:34 GMT","location":"https://vidmoly.me/w/g1ob8a8h5pc8","nel":"{\"report_to\":\"cf-nel\",\"success_fraction\":0.0,\"max_age\":604800}","report-to":"{\"group\":\"cf-nel\",\"max_age\":604800,\"endpoints\":[{\"url\":\"https://a.nel.cloudflare.com/report/v4?s=NHRf6%2BJCPWMVnVvi6%2Bz1FOuJf09NA582XTuP6BDVb3LIpRGFQ2pbMA1RnIszy4%2BZhE8RCPCBHmFMJUPkWlOdseinvF9J7YnWD%2BieHLRWuaY%3D\"}]}","server":"cloudflare","x-content-type-options":"nosniff","x-download-options":"noopen","x-frame-options":"SAMEORIGIN","x-permitted-cross-domain-policies":"none","x-request-id":"GFKJRQKr0B_XEi0StRKB","x-xss-protection":"1; mode=block"}