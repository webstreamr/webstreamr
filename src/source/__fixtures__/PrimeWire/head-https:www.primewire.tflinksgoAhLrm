{"alt-svc":"h3=\":443\"; ma=86400","cache-control":"max-age=0, private, must-revalidate","cf-cache-status":"DYNAMIC","cf-ray":"95fc46ad19eb5c32-FRA","connection":"close","content-type":"text/html; charset=utf-8","cross-origin-window-policy":"deny","date":"Tue, 15 Jul 2025 21:15:35 GMT","location":"https://dood.watch/d/sn87i3ru7plx","nel":"{\"report_to\":\"cf-nel\",\"success_fraction\":0.0,\"max_age\":604800}","report-to":"{\"group\":\"cf-nel\",\"max_age\":604800,\"endpoints\":[{\"url\":\"https://a.nel.cloudflare.com/report/v4?s=x%2BimEPmTu93YQLi7cK%2BINKzJRq7mbm8U%2FKGhNXzrUBEgAn4OMj5AY2dUTOfyinerKldozT%2Fj%2BkDlpUFVLO0gIx2kjLkGcatk1FH2oi4t\"}]}","server":"cloudflare","x-content-type-options":"nosniff","x-download-options":"noopen","x-frame-options":"SAMEORIGIN","x-permitted-cross-domain-policies":"none","x-request-id":"GFKJRVWbP4rTqYARX0gD","x-xss-protection":"1; mode=block"}