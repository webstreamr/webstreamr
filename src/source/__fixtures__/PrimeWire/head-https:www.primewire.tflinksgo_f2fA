{"alt-svc":"h3=\":443\"; ma=86400","cache-control":"max-age=0, private, must-revalidate","cf-cache-status":"DYNAMIC","cf-ray":"95fc474a4e944daa-FRA","connection":"close","content-type":"text/html; charset=utf-8","cross-origin-window-policy":"deny","date":"Tue, 15 Jul 2025 21:16:01 GMT","location":"https://streamtape.com/v/6QLoLq0O8DtObX","nel":"{\"report_to\":\"cf-nel\",\"success_fraction\":0.0,\"max_age\":604800}","report-to":"{\"group\":\"cf-nel\",\"max_age\":604800,\"endpoints\":[{\"url\":\"https://a.nel.cloudflare.com/report/v4?s=VetlSO4eTEkuDVW%2F7uXQuDBeE9TsMzid%2Fvkngehc%2F%2BuKZHo00sRWks3bysCM8EVV4NJCrjiREy4jaJr4RideDjmlk1OHO%2FXzCK5kuZmNE8A%3D\"}]}","server":"cloudflare","x-content-type-options":"nosniff","x-download-options":"noopen","x-frame-options":"SAMEORIGIN","x-permitted-cross-domain-policies":"none","x-request-id":"GFKJSzB71uHjv2MJICgH","x-xss-protection":"1; mode=block"}