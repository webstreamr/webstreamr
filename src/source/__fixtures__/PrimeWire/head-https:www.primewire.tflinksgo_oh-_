{"alt-svc":"h3=\":443\"; ma=86400","cache-control":"max-age=0, private, must-revalidate","cf-cache-status":"DYNAMIC","cf-ray":"95fc46a88fb991e3-FRA","connection":"close","content-type":"text/html; charset=utf-8","cross-origin-window-policy":"deny","date":"Tue, 15 Jul 2025 21:15:35 GMT","location":"https://vidmoly.me/w/awjd3z0t0i9l","nel":"{\"report_to\":\"cf-nel\",\"success_fraction\":0.0,\"max_age\":604800}","report-to":"{\"group\":\"cf-nel\",\"max_age\":604800,\"endpoints\":[{\"url\":\"https://a.nel.cloudflare.com/report/v4?s=v7%2FkU9CoZk0MiSykjeWpG4aSDasJWf%2FzO9QNpzvTnbN5Ff1w8ihZNR4AUxg5pXwTEIagjym2AJtUnMNACSeOSxc568EWXpk8lkhuGZluWME%3D\"}]}","server":"cloudflare","x-content-type-options":"nosniff","x-download-options":"noopen","x-frame-options":"SAMEORIGIN","x-permitted-cross-domain-policies":"none","x-request-id":"GFKJRSoJL1M-uVIQ4ieC","x-xss-protection":"1; mode=block"}