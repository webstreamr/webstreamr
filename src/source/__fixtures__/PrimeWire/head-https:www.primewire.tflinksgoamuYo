{"alt-svc":"h3=\":443\"; ma=86400","cache-control":"max-age=0, private, must-revalidate","cf-cache-status":"DYNAMIC","cf-ray":"95fc468749d41fad-FRA","connection":"close","content-type":"text/html; charset=utf-8","cross-origin-window-policy":"deny","date":"Tue, 15 Jul 2025 21:15:29 GMT","location":"https://voe.sx/y8lrmf2t7jhp","nel":"{\"report_to\":\"cf-nel\",\"success_fraction\":0.0,\"max_age\":604800}","report-to":"{\"group\":\"cf-nel\",\"max_age\":604800,\"endpoints\":[{\"url\":\"https://a.nel.cloudflare.com/report/v4?s=TLOyPNHzk0W2BsQT0HJVbfRQDSzReIO%2FgtOfXuQs43fACZWh5kZ7Ry%2Bda4LM0ZsursBAhMVfbYCxQ6yoSGFKxkP%2FpmlR8yBnlEktGhoW46o%3D\"}]}","server":"cloudflare","x-content-type-options":"nosniff","x-download-options":"noopen","x-frame-options":"SAMEORIGIN","x-permitted-cross-domain-policies":"none","x-request-id":"GFKJQ-yBvaKZeXUStBMB","x-xss-protection":"1; mode=block"}