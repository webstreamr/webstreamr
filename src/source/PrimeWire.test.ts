import { createTestContext } from '../test';
import { FetcherMock, ImdbId } from '../utils';
import { PrimeWire } from './PrimeWire';

const ctx = createTestContext({ en: 'on' });

describe('PrimeWire', () => {
  let handler: PrimeWire;

  beforeEach(() => {
    handler = new PrimeWire(new FetcherMock(`${__dirname}/__fixtures__/PrimeWire`));
  });

  test('handle non-existent movie gracefully', async () => {
    const streams = await handler.handle(ctx, 'movie', new ImdbId('tt12345678', undefined, undefined));
    expect(streams).toHaveLength(0);
  });

  test('handle non-existent episode gracefully', async () => {
    const streams = await handler.handle(ctx, 'series', new ImdbId('tt18546730', 99, 99));
    expect(streams).toHaveLength(0);
  });

  test('handle imdb dead city s2e5', async () => {
    const streams = await handler.handle(ctx, 'series', new ImdbId('tt18546730', 2, 5));
    expect(streams).toMatchSnapshot();
  });

  test('handle el camino', async () => {
    const streams = await handler.handle(ctx, 'series', new ImdbId('tt9243946', undefined, undefined));
    expect(streams).toMatchSnapshot();

    // Should be using the redirectUrlCache
    const streams2 = await handler.handle(ctx, 'series', new ImdbId('tt9243946', undefined, undefined));
    expect(streams2).toMatchSnapshot();
  });
});
