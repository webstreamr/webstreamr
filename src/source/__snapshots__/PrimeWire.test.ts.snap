// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PrimeWire handle el camino 1`] = `
[
  {
    "countryCode": "en",
    "url": "https://mixdrop.ag/f/4njk6691b1k60d",
  },
  {
    "countryCode": "en",
    "url": "https://mixdrop.ag/f/2bnmi",
  },
  {
    "countryCode": "en",
    "url": "https://mixdrop.ag/f/ql0qrxnjb89nx8",
  },
  {
    "countryCode": "en",
    "url": "https://voe.sx/paqi7qqku3zi",
  },
  {
    "countryCode": "en",
    "url": "https://voe.sx/ysimllytd3zo",
  },
  {
    "countryCode": "en",
    "url": "https://voe.sx/xwhvakkexuob",
  },
  {
    "countryCode": "en",
    "url": "https://filelions.to/f/wayeo7l1cw7d",
  },
  {
    "countryCode": "en",
    "url": "https://filelions.to/f/bv4b7tl9u1ij",
  },
  {
    "countryCode": "en",
    "url": "https://streamwish.to/09q73jf6t6jw",
  },
  {
    "countryCode": "en",
    "url": "https://dood.watch/d/05fm25orfeux",
  },
  {
    "countryCode": "en",
    "url": "https://dood.watch/d/9lw62yyosxc0",
  },
  {
    "countryCode": "en",
    "url": "https://dood.watch/d/j2pa25q11pdu",
  },
  {
    "countryCode": "en",
    "url": "https://streamtape.com/v/qMVMPa49M1hz8BR/El.Camino.A.Breaking.Bad.Movie.2019.1080p.WEBRip.x264_YTS.LT.mp4",
  },
  {
    "countryCode": "en",
    "url": "https://streamtape.com/v/6QLoLq0O8DtObX",
  },
]
`;

exports[`PrimeWire handle el camino 2`] = `
[
  {
    "countryCode": "en",
    "url": "https://mixdrop.ag/f/4njk6691b1k60d",
  },
  {
    "countryCode": "en",
    "url": "https://mixdrop.ag/f/2bnmi",
  },
  {
    "countryCode": "en",
    "url": "https://mixdrop.ag/f/ql0qrxnjb89nx8",
  },
  {
    "countryCode": "en",
    "url": "https://voe.sx/paqi7qqku3zi",
  },
  {
    "countryCode": "en",
    "url": "https://voe.sx/ysimllytd3zo",
  },
  {
    "countryCode": "en",
    "url": "https://voe.sx/xwhvakkexuob",
  },
  {
    "countryCode": "en",
    "url": "https://filelions.to/f/wayeo7l1cw7d",
  },
  {
    "countryCode": "en",
    "url": "https://filelions.to/f/bv4b7tl9u1ij",
  },
  {
    "countryCode": "en",
    "url": "https://streamwish.to/09q73jf6t6jw",
  },
  {
    "countryCode": "en",
    "url": "https://dood.watch/d/05fm25orfeux",
  },
  {
    "countryCode": "en",
    "url": "https://dood.watch/d/9lw62yyosxc0",
  },
  {
    "countryCode": "en",
    "url": "https://dood.watch/d/j2pa25q11pdu",
  },
  {
    "countryCode": "en",
    "url": "https://streamtape.com/v/qMVMPa49M1hz8BR/El.Camino.A.Breaking.Bad.Movie.2019.1080p.WEBRip.x264_YTS.LT.mp4",
  },
  {
    "countryCode": "en",
    "url": "https://streamtape.com/v/6QLoLq0O8DtObX",
  },
]
`;

exports[`PrimeWire handle imdb dead city s2e5 1`] = `
[
  {
    "countryCode": "en",
    "url": "https://mixdrop.ag/f/q1pz4k94uxxkgvo",
  },
  {
    "countryCode": "en",
    "url": "https://mixdrop.ag/f/nlpj3q3ran17rv",
  },
  {
    "countryCode": "en",
    "url": "https://mixdrop.ag/f/9wvz9gp3b3q1llk",
  },
  {
    "countryCode": "en",
    "url": "https://voe.sx/grsvpdbxvw3e",
  },
  {
    "countryCode": "en",
    "url": "https://voe.sx/te40pmzra8du",
  },
  {
    "countryCode": "en",
    "url": "https://voe.sx/y8lrmf2t7jhp",
  },
  {
    "countryCode": "en",
    "url": "https://filelions.to/f/eqmzrkv5kf5s",
  },
  {
    "countryCode": "en",
    "url": "https://filelions.to/f/0i0j0qz5o478",
  },
  {
    "countryCode": "en",
    "url": "https://filelions.to/f/odssr9dyjx5n",
  },
  {
    "countryCode": "en",
    "url": "https://filelions.to/f/fpru8wthtsf4",
  },
  {
    "countryCode": "en",
    "url": "https://streamwish.to/jult1lcvqjlt",
  },
  {
    "countryCode": "en",
    "url": "https://streamwish.to/h9p9qqgpq2lw",
  },
  {
    "countryCode": "en",
    "url": "https://streamwish.to/s986lkdy8mb5",
  },
  {
    "countryCode": "en",
    "url": "https://vidmoly.me/w/g1ob8a8h5pc8",
  },
  {
    "countryCode": "en",
    "url": "http://vidmoly.me/w/46bi0sec0966",
  },
  {
    "countryCode": "en",
    "url": "https://vidmoly.me/w/awjd3z0t0i9l",
  },
  {
    "countryCode": "en",
    "url": "http://vidmoly.me/w/okcwblnaqw3t",
  },
  {
    "countryCode": "en",
    "url": "https://dood.watch/d/sn87i3ru7plx",
  },
  {
    "countryCode": "en",
    "url": "https://dood.watch/d/8odyvc9bre16",
  },
  {
    "countryCode": "en",
    "url": "https://dood.watch/d/1qfuy5yiq4r6",
  },
  {
    "countryCode": "en",
    "url": "https://dood.watch/d/reslxv9siz41",
  },
  {
    "countryCode": "en",
    "url": "https://streamplay.to/qyn6mtz6w3sr",
  },
  {
    "countryCode": "en",
    "url": "https://streamplay.to/m1j7zix6ny8u",
  },
  {
    "countryCode": "en",
    "url": "https://streamplay.to/yrkzvpayg27o",
  },
  {
    "countryCode": "en",
    "url": "https://streamtape.com/v/Ap0GJoXglaFbqK/The.Walking.Dead.Dead.City.S02E05.The.Bird.Always.Knows.720p.AMZN.WEB-DL.DDP5.1.H.264-NTb.mkv.mp4",
  },
  {
    "countryCode": "en",
    "url": "https://streamtape.com/v/g3dQvLvy9QsJqo/The.Walking.Dead.Dead.City.S02E05.The.Bird.Always.Knows.720p.AMZN.WEB-DL.DDP5.1.H.264-NTb.mkv.mp4",
  },
  {
    "countryCode": "en",
    "url": "https://filemoon.sx/d/8wllnog0cyl7",
  },
  {
    "countryCode": "en",
    "url": "https://filemoon.sx/d/ftowmfa7f0ge/The.Walking.Dead.Dead.City.S02E05.The.Bird.Always.Knows.720p.AMZN.WEB-DL.DDP5.1.H.264-NTb.mkv",
  },
  {
    "countryCode": "en",
    "url": "https://dropload.io/ubaaot4b6pfw",
  },
  {
    "countryCode": "en",
    "url": "https://dropload.io/boj6hx5v0j96",
  },
  {
    "countryCode": "en",
    "url": "https://savefiles.com/dkq2hy8nt1s8",
  },
  {
    "countryCode": "en",
    "url": "https://savefiles.com/i0bj9gazequy",
  },
  {
    "countryCode": "en",
    "url": "https://savefiles.com/tsbt84uhcvnv",
  },
  {
    "countryCode": "en",
    "url": "https://luluvdoo.com/nrcy0m01ce37",
  },
  {
    "countryCode": "en",
    "url": "https://luluvdoo.com/2lyf7xx54140",
  },
  {
    "countryCode": "en",
    "url": "https://luluvdoo.com/d/au00sj11vqku",
  },
  {
    "countryCode": "en",
    "url": "https://filegram.to/6q8gctbf79gc",
  },
  {
    "countryCode": "en",
    "url": "https://filegram.to/erapbfgcdguq",
  },
  {
    "countryCode": "en",
    "url": "https://filegram.to/6yuyz0dgqu05",
  },
  {
    "countryCode": "en",
    "url": "https://bigwarp.cc/4tb37obzt6re",
  },
  {
    "countryCode": "en",
    "url": "https://bigwarp.cc/aazhpuex9ppj",
  },
  {
    "countryCode": "en",
    "url": "https://bigwarp.cc/u66q6mybilfa",
  },
]
`;
